### ISSUE TITLE
JSDoc: inferred type from class instance does not respect optional constructor fields

### ISSUE THREAD

**Issue by theetrain** [microsoft/TypeScript#57416]
### 🔎 Search Terms

Class, constructor, jsdoc, inferred, types

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about [derived class property initializers](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-do-my-derived-class-property-initializers-overwrite-values-set-in-the-base-class-constructor).

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.3&filetype=js#code/MYGwhgzhAECCB2BLAtmE0DeAoavoHoAqQnPaQ6AAQAcwAnMZTDaeRgUwH4AuaCAFzqJ4Ac2gBfcdABmidiAAmEUnkL4V0YAHt4AugFdg-LXQAUs+UoCUmDWX4ALRBAB0bZO2gBeGXMWv3dg1xLBCsbV1+aBBEHW9WdgB3OCRUEFMWQN4AcgAhLQAjbPErcJ0BaGpweCifeCSUlDQMkqwsIgoaekZMPWERKUDydWl9eCNY+GgRdn4AOQ5TQJtsMgiILRB2FxAtESWOUrCZ+cWYnTdDrBOFj1MqsBrLjxt8fD4HLX1FaEc6LWS-AAntRPOw6P86EA

### 💻 Code

```js
class Animal {
    /**
     * @param {{ name?: string }} fields
     */
    constructor(fields) {
        this.name = fields.name
    }
}

const lion = new Animal({ name: 'Bob'})
const plant = new Animal({})

/** @param {string} name */
function getName(name) {
    console.log(name)
}

getName(lion.name)
getName(plant.name) // should throw type error
```

### 🙁 Actual behavior

- My class instance `const plant = new Animal({})` should have its `plant.name` property type inferred as `undefined` since it wasn't instantiated.
- When passing `plant.name` as a parameter to my `getName()` function, it should expect its first parameter typed as `string`, but doesn't throw an error

### 🙂 Expected behavior

Class instances should have their public field types inferred from the constructor:

```js
const lion = new Animal({ name: 'Bob'}) // `lion.name` should be type `string`
const plant = new Animal({}) // `plant.name` should be type `undefined`
```

**Comment by MartinJohns** [comment]
This is working as intended.

> * My class instance `const plant = new Animal({})` should have its `plant.name` property type inferred as `undefined` since it wasn't instantiated.

Types are based on the available type information, not runtime values. The type information available is this part: `{ name?: string }` - so it's always an optional `string`. If you want to have different types depending on the type of the argument, then you need to use a generic.

> * When passing `plant.name` as a parameter to my `getName()` function, it should expect its first parameter typed as `string`, but doesn't throw an error

Because you have `strictNullChecks` disabled. Without this flag the type `undefined` is basically not existing. The type of `name` is `string` in this case.

**Comment by theetrain** [comment]
Thanks for explaining. Moving forward I'll either use generics or type assertions where appropriate.
